core/new-gui/src/app/workspace/types/shared-editing.interface.ts

```diff
@@ -0,0 +1,125 @@
+import * as Y from "yjs";
+import * as _ from "lodash";
+
+export type YTextify<T> = T extends string ? Y.Text : T;
+export type YArrayify<T> = T extends Array<any> ? Y.Array<any> : T;
+
+/**
+ * <code>YType<T></code> is the yjs-object version of a normal js object with type <code>T</code>.
+ *
+ * Additionally, <code>YType</code> preserves <code>keyof</code> requirements from the original object.
+ *
+ * <code>toJSON()</code> converts the <code>YType</code> back to a normal js object.
+ */
+export type YType<T> = Omit<Y.AbstractType<any>, "get" | "set" | "has" | "toJSON"> & {
+  get<TKey extends keyof T>(key: TKey): YArrayify<YTextify<T[TKey]>>;
+  set<TKey extends keyof T>(key: TKey, value: YArrayify<YTextify<T[TKey]>>): void;
+  has<TKey extends keyof T>(key: TKey): boolean;
+  toJSON(): T;
+};
+
+/** Creates a <code>YType</code> given a normal object. Returns either a <code>YType</code>,
+ *  or the original object if it is a primitive type other than string, because string will be converted to
+ *  <code>Y.Text</code>.
+ *  @param obj: a normal object, could be either a string, an array, or a complicated object with its own attributes.
+ */
+export function createYTypeFromObject<T extends object>(obj: T): YType<T> {
+  const objType = obj.constructor.name;
+  if (objType === "String") {
+    return new Y.Text(obj as unknown as string) as unknown as YType<T>;
+  } else if (objType === "Array") {
+    const yArray = new Y.Array();
+    // Create YType for each array item and push
+    for (const item of obj as any) {
+      yArray.push([createYTypeFromObject(item) as unknown]);
+    }
+    return yArray as unknown as YType<T>;
+  } else if (objType === "Object") {
+    // return new
+    const yMap = new Y.Map();
+    Object.keys(obj).forEach((k: string) => {
+      const value = obj[k as keyof T] as any;
+      if (value !== undefined) {
+        yMap.set(k, tryToCreateYTypeFromObject(value));
+      }
+    });
+    return yMap as unknown as YType<T>;
+  } else {
+    throw TypeError(`Cannot create YType from ${objType}!`);
```

Comment:
Done. The new code will _NOT_ accept primitive types from user-code as input on compilation, but it will accept primitive types if they are recursively and internally passed to this function during JS runtime.